enterprise-best-practices/q-coding-guidelines.md

```diff
@@ -0,0 +1,605 @@
+Q Coding Guidelines
+===================
+
+This is a collection of guidelines and best practices for `q`.
+
+Introduction
+------------
+
+Why have code conventions ? It should be clear that readability of code
+is essential for code maintainability. This applies to all software
+languages, but becomes even more important in languages with such high
+density of notation as `q`.
+
+### Acknowledgements
+
+This initial release of these guidelines are based on notes and training
+materials authored by Jeff Borror, Charlie Skelton's (kx.com) guidelines
+and Stevan Apter's [Remarks on
+Style](http://www.nsl.com/papers/style.pdf).
+
+### Motivation
+
+-   Code is:
+    -   written once
+    -   modified 10 times
+    -   read 100 times
+-   Write for those who come after you
+
+### Q coding goals
+
+-   short vs simple:
+    -   Simple problems usually have simple solutions in q
+    -   Fewer operators often but not always means simpler and better
+    -   Code count should not be the primary metric of q code goodness
+-   Q coding strategy:
+    -   *Find a solution that works, then look for simpler solutions. If
+        performance is critical, test and pick the solution with the
+        quickest execution.*
+-   Go with the flow:
+    -   Use data constructs and idioms of `q`
+    -   Do not impose constructs from other languages
+    -   Use vector operations whenever reasonable
+    -   Do not make your q code look like: C, Java, Perl, SQL, or ...
+
+File Names
+----------
+
+-   suffixes: `q` scripts must have the `.q` extension
+-   names: `q` scripts should use alphanumeric names without whitespace
+
+File Organisation
+-----------------
+
+-   beginning comments: use a comment block that lists Perforce keywords
+    for filename, version, date and author info as well as a copyright
+    notice
+-   use [QDoc](QDoc) for generating automated docs for your code
```

Comment:
Broken link (currently). Did you mean to point to [jasaraj/q-doc](https://github.com/jasraj/q-doc)?